### PR TITLE
Extract docs

### PR DESCRIPTION
**Pull request by tmlayton**
## Todo
- [ ] Fix squished description column
- [x] Update output paths to match [the new taxonomies](https://docs.google.com/document/d/1n00ATZhsR6UgZ5j0CxuFfCWbj3UyUtLFkYdE4kJSB6g/edit#)
- [x] Replace Github relative links with the Shopify.dev equivalent
- [x] Fix call stack error with View component
- [x] Fix Responsive props - resolving as an "UndocumentedType", maybe because it is an interface that is part of a union
- [ ] Preserve unique locals (e.g. `type Size` in BlockSpacing, Text, and Layout are all different). Currently docs for Layout are incorrect and listing the string literal values for BlockSpacing.
- [ ] Add generic arguments to interface titles
- [ ] Generate both READMEs and shopify.dev docs
- [ ] Parse `extends` - in progress
- [x] Handle more complex types like generics or literal interfaces (e.g. Text `role`)
- [x] Parse tsdoc tags (e.g. `@default`, etc) - https://github.com/Shopify/argo-checkout/pull/73
- [x] Fix `void` return for functions
- [x] Parse `/examples/name.example.[ts|*]` files
- [x] Parse `/content` folder to additonal markdown content
- [x] Use markdown tables OR properly parse basic markdown in table content (e.g. [markdown links](#links))
- [x] Fix overflow formatting on table
- [ ] Metadata file?
- [ ] Publish as versions?
- [ ] Refactor all of this into composable parts to share with Argo Admin 😂 

## Content todo

- [ ] Write examples for vanilla JS and React
- [ ] Pull markdown formatting out and place in content files (e.g. CalloutBanner "Guidelines")
- [ ] Update all doc comments to inline content. Flatten any lists or move to `/content` markdown files.
- [ ] Move any doc comments to interfaces. Comments on `type`’s used in interfaces are not parsed.

### CODE REVIEW

**Comment by iainmcampbell** [comment]
CI is failing on what appears to be a lack of dom types (https://github.com/ReactiveX/rxjs/issues/1986#issuecomment-268375547)

…however, we already include the `dom` library in our `compilerOptions` in `tsconfig.base.json` 🤔

**Comment by tmlayton** [comment]
Type-check failure is because of the typescript compiler configuration does not include the DOM lib (because its intended to be a web worker) so it doesn’t have a type for `document`.

**Comment by tmlayton** [comment]
It looks like the TS config for the argo-checkout package is being used, which appears to overwrite the array of `lib`'s when using a base config instead of merging them. When I swap `WebWorker` with `dom` the error goes away but I get another error related to the global declaration for `shopify`, it’s also not the right thing to do. I tried using both the web worker and dom libs but it throws a bunch of conflict errors.

**Comment by iainmcampbell** [comment]
I want to merge this in cause it's becoming a pain to rebase. It's also modular enough that Admin and Checkout can build on this separately.

The scripts are still WIP but if you wanna test them, make sure you have shopify-dev cloned and run `yarn docs:admin` or `yarn docs:checkout`.

**Comment by tmlayton** [comment]
Not sure how I feel about the `skipLibCheck` in the base config, we should be able to instead move the scripts to their own area with their own tsconfig which includes the dom lib.

Otherwise I’m fine with merging. If you do so I just ask that you move the outstanding todos to their own issues.

**Review by vividviolet — COMMENTED** [review]
I couldn't get very far for testing unfortunately

File: scripts/typedoc/shopify-dev-renderer/index.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import {Paths} from '../types';
+
+import {gettingStarted} from './getting-started';
+import {extensionPoints} from './extension-points';
+import {components} from './components';
+
+export {extensionPoints, components, gettingStarted};
+
+export function renderForShopifyDev(paths: Paths) {
```
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/typedoc/shopify-dev-renderer/index.ts:9]: hm, you might need to `yarn install` in checkout-web?

File: tsconfig.json [hunk 2]
```diff
@@ -10,6 +10,7 @@
     {"path": "./packages/argo-webpack-hot-client"},
     {"path": "./packages/argo-run"},
     {"path": "./packages/argo-admin"},
-    {"path": "./packages/argo-admin-react"}
+    {"path": "./packages/argo-admin-react"},
+    {"path": "./scripts"}
```
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply tsconfig.json:14]: sure? doesn’t fix the issue though, getting the same `error TS2304: Cannot find name 'HTMLDocument'`.

File: scripts/generate-docs-admin.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  packages: {
+    JavaScript: './packages/argo-admin',
+    React: './packages/argo-admin-react',
+  },
+  outputRoot: '../shopify-dev/content/tools/argo-admin',
+  shopifyDevUrl: '/tools/argo-admin',
+};
+
+components(paths);
+extensionPoints(paths);
```
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/generate-docs-admin.ts:14]: fixed this, sorry!

File: scripts/typedoc/shopify-dev-renderer/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/typedoc/shopify-dev-renderer/index.ts:9]: I don't think we can/should clone repos automatically. i'll write a readme with these prereq steps.

File: scripts/generate-docs-admin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (vividviolet) [reply scripts/generate-docs-admin.ts:14]: I just tried it again and it doesn't seem to work still. First time I run it I get:

```
Error: ENOENT: no such file or directory, mkdir '/Users/trishta/src/github.com/Shopify/shopify-dev/content/tools/argo-admin/extension-points'
    at Object.mkdirSync (node:fs:1038:3)
    at extensionPoints (/Users/trishta/src/github.com/Shopify/argo-checkout/scripts/typedoc/shopify-dev-renderer/extension-points.ts:52:8)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  errno: -2,
  syscall: 'mkdir',
  code: 'ENOENT',
  path: '/Users/trishta/src/github.com/Shopify/shopify-dev/content/tools/argo-admin/extension-points'
}
```
Second time I run it it works aside from the warnings:

```
TSIntersectionType is unhandled.
Can’t resolve export type `DestructableAction` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `DestructableAction` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `DisableableAction` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `DestructableAction` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `DestructableAction` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/argo-checkout/packages/argo-admin/src/components/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
✨  Done in 1.66s.
``

File: scripts/generate-docs-admin.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import {components, extensionPoints} from './typedoc/shopify-dev-renderer';
+
+const paths = {
+  inputRoot: './packages/argo-admin',
+  packages: {
+    JavaScript: './packages/argo-admin',
+    React: './packages/argo-admin-react',
+  },
+  outputRoot: '../shopify-dev/content/tools/argo-admin',
```
Comment (vividviolet) [inline scripts/generate-docs-admin.ts:9]: Should we not put these docs in the same place as Checkout like `docs/admin-extensions`?

File: scripts/typedoc/shopify-dev-renderer/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (vividviolet) [reply scripts/typedoc/shopify-dev-renderer/index.ts:9]: I'm still seeing the type errors, are these expected?
```
TSTypeQuery is unhandled.
ArrowFunctionExpression is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSIndexedAccessType is unhandled.
TSConditionalType is unhandled.
TSConditionalType is unhandled.
TSConditionalType is unhandled.
TSTypeLiteral is unhandled.
TSTypeLiteral is unhandled.
TSTypeLiteral is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTypeLiteral is unhandled.
TSTypeLiteral is unhandled.
TSTypeLiteral is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
TSTupleType is unhandled.
Can’t resolve export type `Record` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `AllComponents` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Record` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `StatefulRemoteSubscribable` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Promise` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `AllComponents` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `AllowedComponents` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `RemoteRoot` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Input` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Input` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Input` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Input` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
Can’t resolve export type `Partial` in /Users/trishta/src/github.com/Shopify/checkout-web/packages/argo-checkout/src/index.ts. Maybe it’s not exported from the component index or imported from a remote package.
```

**Review by vividviolet — COMMENTED** [review]
I finally got the scripts working but due to local issues I can't actually see the docs on shopify-dev so I couldn't fully top hat. The files look great though aside from the warnings

File: scripts/generate-docs-admin.ts (hunk 4 above)
Comment (iainmcampbell) [reply scripts/generate-docs-admin.ts:9]: probably yes, haven't finalized this yet. not concerned about this right now.

File: scripts/typedoc/shopify-dev-renderer/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/typedoc/shopify-dev-renderer/index.ts:9]: yes, they're expected, we are throwing these https://github.com/Shopify/argo-checkout/pull/70/files/0202003c242c338e73c97d1daea03cf24ff4e908#diff-b08df2e37da7db49e6cf3a91d88d8c459dde06173eafd476d770e275c5baf7e8R569-R571

[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/typedoc/shopify-dev-renderer/index.ts:9]: this script is still WIP! I will need to handle some of these for the extension points, others we can safely ignore

File: scripts/generate-docs-admin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (iainmcampbell) [reply scripts/generate-docs-admin.ts:14]: it turns out node can't create directories recursively. i've fixed my fix for the first error.

**Comment by iainmcampbell** [comment]
🎉🎉🎉
Screen Shot 2021-05-05 at 3 26 21 PM

**Review by vividviolet — APPROVED** [review]
Works great 👍

**Pull request merged**